File: cpp/1358.number-of-substrings-containing-all-three-characters.cpp
class Solution {
public:
    int numberOfSubstrings(string s) {
        const int n = s.size();

        int cnt[] = {0, 0, 0};
        int win = 0, res = 0;
        int i = 0;

        for (int j = 0; j < n; j++) {
            cnt[s[j] - 'a']++;

            while (cnt[0] && cnt[1] && cnt[2]) {
                win++;
                cnt[s[i] - 'a']--;
                i += 1;
            }

            res += win;
        }

        return res;
    }
};

File: python/1358.number-of-substrings-containing-all-three-characters.py
class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        n = len(s)
        a = ord("a")
        win = [0,0,0]
        res = 0

        l = 0
        for r in range(n):
            win[ord(s[r]) - a] += 1

            while win[0] and win[1] and win[2]:
                res += n - r # add the rest of string l times

                win[ord(s[l]) - a] -= 1
                l += 1

        return res
        